apps/prairielearn/src/lib/server-jobs-2.ts

```diff
@@ -0,0 +1,184 @@
+import socketServer = require('./socket-server');
+
+const sql = loadSqlEquiv(__filename);
+
+interface CreateServerJobOptions {
+  courseId?: string;
+  courseInstanceId?: string;
+  courseRequestId?: string;
+  assessmentId?: string;
+  userId?: string;
+  authnUserId?: string;
+  type: string;
+  description: string;
+}
+
+interface ServerJobExecOptions {
+  cwd?: string;
+  env?: NodeJS.ProcessEnv;
+}
+
+type ServerJobContext = Omit<ServerJob, 'execute' | 'executeInBackground'>;
+type JobExecutionFunction = (context: ServerJobContext) => Promise<void>;
+
+class ServerJob {
+  public jobSequenceId: string;
+  public jobId: string;
+  private started = false;
+  private finished = false;
+  public output = '';
+
+  constructor(jobSequenceId: string, jobId: string) {
+    this.jobSequenceId = jobSequenceId;
+    this.jobId = jobId;
+  }
+
+  error(msg: string) {
+    this.addToOutput(msg + '\n');
+  }
+
+  warn(msg: string) {
+    this.addToOutput(msg + '\n');
+  }
+
+  info(msg: string) {
+    this.addToOutput(msg + '\n');
+  }
+
+  verbose(msg: string) {
+    this.addToOutput(msg + '\n');
+  }
+
+  async exec(file: string, args: string[] = [], options: ServerJobExecOptions = {}): Promise<void> {
+    const msg = `$ ${file} ${args.join(' ')}\n`;
+    this.addToOutput(chalk.greenBright(msg));
+    const proc2 = execa(file, args, {
+      ...options,
+      reject: false,
+      all: true,
+    });
+
+    proc2.all?.setEncoding('utf-8');
+    proc2.all?.on('data', (data) => {
+      this.addToOutput(data);
+    });
+
+    const result = await proc2;
+    if (result.exitCode !== 0) {
+      throw new Error(`Process exited with code ${result.exitCode}`);
+    }
+  }
+
+  /**
+   * Runs the job sequence and returns a Promise that resolves when the job
+   * sequence has completed. The returned promise will not reject if the job
+   * sequence fails.
+   */
+  async execute(fn: JobExecutionFunction): Promise<void> {
+    if (this.started) {
+      throw new Error('ServerJob already started');
+    }
+    this.started = true;
+
+    try {
+      await fn(this);
+      await this.finish();
+    } catch (err) {
+      try {
+        await this.finish(err);
+      } catch (err) {
+        logger.error(`Error failing job ${this.jobId}`, err);
+        Sentry.captureException(err);
+      }
+    }
+  }
+
+  /**
+   * Identical to {@link execute}, except it doesn't return a Promise. This is
+   * just used as a hint to the caller that they don't need to wait for the job
+   * to finish. Useful for tools like TypeScript and ESLint that ensure that
+   * promises are awaited.
+   */
+  executeInBackground(fn: JobExecutionFunction): void {
+    this.execute(fn);
+  }
+
+  private addToOutput(msg: string) {
+    this.output += msg;
+    const ansiUp = new AnsiUp();
+    const ansifiedOutput = ansiUp.ansi_to_html(this.output);
+    socketServer.io
+      .to('job-' + this.jobId)
+      .emit('change:output', { job_id: this.jobId, output: ansifiedOutput });
```

Comment:
I haven't yet implemented the thing where we only send deltas. Thankfully, it seems like `ansi_up` can work in a streaming manner, so this shouldn't be too hard to do in another PR! We just need to retain a single `AnsiUp` instead for each job.

Reply:
Oh cool, that will be very nice to have!